File: exercism/elixir/hamming/lib/hamming.ex
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2), do: {:error, "strands must be of equal length"}
  def hamming_distance(strand1, strand2), do: do_hamming_distance(strand1, strand2, 0)

  defp do_hamming_distance([], [], diff), do: {:ok, diff}
  defp do_hamming_distance([h1 | t1], [h2|t2], diff) do
    do_hamming_distance(t1, t2, (if h1 != h2, do: diff + 1, else: diff))
  end
end

File: exercism/lua/hamming/hamming.lua
local Hamming = {}

function Hamming.compute(a,b)
    local la = #a
    if la ~= #b then
        return -1
    end

    local diff = 0
    for i=1,la+1 do
        if a:byte(i) ~= b:byte(i) then diff = diff + 1 end
    end
    return diff
end

return Hamming
